classes/itemlist.php

```diff
@@ -791,41 +791,58 @@ public function display_relations_table() {
     }
 
     /**
-     * Adds elements to an array starting from initial conditions.
-     *
-     * Conditions are array('hidden' => 0) OR array('reserved' => 0)
-     * Add to $nodelist all the children that will inherit the parent condition
+     * Get the recursive list of children of a specific item.
+     * This list counts children and children of children for as much generation as it is.
      *
-     * @param integer $nodelist
-     * @param array $additionalcondition
-     * @return void
+     * @param int $oldestparent
+     * @param array $where
+     * @return object $childrenitems
      */
-    public function add_child_node(&$nodelist, $additionalcondition) {
+    public function item_get_children($oldestparent=null, $where=null) {
         global $DB;
 
-        if (!is_array($additionalcondition)) {
-            $a = 'add_child_node';
+        if (empty($oldestparent)) {
+            $oldestparent = $this->itemid;
+        }
+
+        if (empty($where)) {
+            $where = array();
+        }
+
+        if (!is_array($where)) {
+            $a = 'item_get_children';
             print_error('arrayexpected', 'mod_surveypro', null, $a);
         }
 
-        $itemid = end($nodelist);
-        $where = array('parentid' => $itemid) + $additionalcondition;
-        if ($childrenitems = $DB->get_records('surveypro_item', $where, 'sortindex', 'id, sortindex')) {
-            foreach ($childrenitems as $childitem) {
-                $nodelist[$childitem->sortindex] = (int)$childitem->id;
-                $this->add_child_node($nodelist, $additionalcondition);
+        $idscontainer = array($oldestparent);
+
+        // Lets start populating the list of items to return.
+        $childrenitems = $DB->get_records('surveypro_item', array('id' => $oldestparent), 'sortindex', 'id, parentid, sortindex');
+
+        $childid = $oldestparent;
+        $i = 1;
+        do {
```

Comment:
Surely here I'd have tried some implementation using recursiveness, instead of that dynamic (growing) looping. But if it's working for you (tried with > 2 levels trees) I don't have much to object. Maybe, at some point, we could make the the ìtemlist` class aware of all items and then we could save some database queries, just looking for the tree in PHP. But I imagine it's not super-critical.

Reply:
> Surely here I'd have tried some implementation using recursiveness
With this PR I am replacing the old recursive routine. I find it like to kill an ant with a bazooka

> Maybe, at some point, we could make the the ìtemlist` class aware of all items and then we could save some database queries
I would be magnificent!